### PR TITLE
feat: Add Sources in Events tree

### PR DESCRIPTION
**Pull request by evidolob**
Current state:
Screenshot 2021-04-13 at 11 29 24

To be compatible with vscode-knative we need to have rest of evet subtree items `Brokers`, `Channels`, `Subscriptions` and `Triggers` as sink node is point to them

### CODE REVIEW

**Comment by joshuawilson** [comment]
Nicely done. A lot added here.

**Comment by lstocchi** [comment]
@evidolob i'm playing with knative eventing and using your PR to visualize the sources. I found that a source can also have an empty sink so we should cover that as well.
image

You should be able to replicate it by following the example here https://knative.dev/docs/eventing/broker/

**Comment by joshuawilson** [comment]
> @evidolob i'm playing with knative eventing and using your PR to visualize the sources. I found that a source can also have an empty sink so we should cover that as well.
> image
> 
> You should be able to replicate it by following the example here https://knative.dev/docs/eventing/broker/

I do not see how a Source can be missing a Sink. It is required. If you run the `kn` command for creating a source it will error without the sink.

**Comment by lstocchi** [comment]
> I do not see how a Source can be missing a Sink. It is required. If you run the `kn` command for creating a source it will error without the sink.
> 
> If you mean that it has the sink specified but it can't find it, it throws an error for that too.

I really don't know what's the point of adding these comments (this is the second time). To show that I said a bullshit and Yevhen (or whoever) doesn't need to verify it? It would be more productive to verify/replicate/test what i wrote and tell me if what i face is 
just me or not. 

Users can also use kubectl to create a source and the tutorial does it this way. If you create a source before the sink (it is possible with kubectl) or the source fails (as it did for me just by following the tutorial above), the plugin displays an empty label because the sink is an empty object and the plugin doesn't handle this situation.

```
{
    "kind": "PingSource",
    "apiVersion": "sources.knative.dev/v1alpha2",
    "metadata": {
        "name": "test-ping-source",
        "namespace": "default",
        ....
        "annotations": {
            "kubectl.kubernetes.io/last-applied-configuration": "{\"apiVersion\":\"sources.knative.dev/v1beta2\",\"kind\":\"PingSource\",\"metadata\":{\"annotations\":{},\"name\":\"test-ping-source\",\"namespace\":\"default\"},\"spec\":{\"contentType\":\"application/json\",\"data\":\"{\\\"message\\\": \\\"Hello world!\\\"}\",\"schedule\":\"*/1 * * * *\",\"sink\":{\"ref\":{\"apiVersion\":\"eventing.knative.dev/v1\",\"kind\":\"Broker\",\"name\":\"default\"}}}}\n",
            "pingsources.sources.knative.dev/v1beta2-spec": "{\"sink\":{\"ref\":{\"kind\":\"Broker\",\"name\":\"default\",\"apiVersion\":\"eventing.knative.dev/v1\"}},\"schedule\":\"*/1 * * * *\",\"contentType\":\"application/json\",\"data\":\"{\\\"message\\\": \\\"Hello world!\\\"}\"}",
            "sources.knative.dev/creator": "minikube-user",
            "sources.knative.dev/lastModifier": "minikube-user"
        },
        ....
    },
    "spec": {
        "sink": {}, <--- empty
        "schedule": "* * * * *"
    }
```

**Comment by joshuawilson** [comment]
I misunderstood your previous statement and thought you were saying you could create a source without a `sink`. I was pointing out that you can not do that intentionally. However there are bugs that allow it to happen. I did not see a missing `sink` by reading the doc, but as you say it fails when you run it. There is a bug in the doc. The pingSource example is using `sources.knative.dev/v1beta2` and that does not work with the latest version of `kn`.  If you run it with `kn source ping create` you should not get that error. 

For @evidolob I suggest that you catch it when the sink is not found and display `Sink Not Found` with no icon.

**Comment by evidolob** [comment]
@lstocchi Now tree show error if sink name is missing:
Screenshot 2021-06-02 at 10 51 07

**Review by jeffmaury — APPROVED** [review]
LGTM

**Review by lstocchi — CHANGES_REQUESTED** [review]
Everything looks good. Just a doubt on the services node.

File: src/main/java/com/redhat/devtools/intellij/knative/tree/KnTreeStructure.java [hunk 1]
```diff
@@ -128,7 +172,10 @@ protected Config loadConfig() {
             Kn kn = element.getRootNode().getKn();
             kn.getServicesList().stream().forEach(it -> services.add(new KnServiceNode(element.getRootNode(), element, getService(kn, it))));
         } catch (IOException e) {
-            services.add(new MessageNode(element.getRootNode(), element, "Failed to load services"));
+            services.add(new MessageNode<>(element.getRootNode(), element, "Failed to load services"));
+        }
+        if (services.isEmpty()) {
+            services.add(new MessageNode<>(element.getRootNode(), element, "No Services Found"));
```
[the remark this answers is not in the document]
Comment (evidolob) [reply src/main/java/com/redhat/devtools/intellij/knative/tree/KnTreeStructure.java:178]: Not sure what I mean when add that `No Services Found` node, maybe mimic to vscode-knative. But I agree that we need to be consistent on tree behaviour, going to delete that.

**Review by lstocchi — APPROVED** [review]
LGTM :rocket: Thanks!!

**Pull request merged**